driver-core/src/main/com/mongodb/operation/MixedBulkWriteOperation.java

```diff
@@ -479,18 +480,22 @@ private boolean shouldAcknowledge(final BulkWriteBatch batch, final SessionConte
             @Override
             public void onResult(final BsonDocument result, final Throwable t) {
                 if (t != null) {
-                    if (isSecondAttempt || !shouldAttemptToRetryWrite(retryWrites, t)) {
-                        if (retryWrites && !isSecondAttempt) {
-                            logUnableToRetry(batch.getPayload().getPayloadType().toString(), t);
-                        }
-                        if (t instanceof MongoWriteConcernWithResponseException) {
-                            addBatchResult((BsonDocument) ((MongoWriteConcernWithResponseException) t).getResponse(), binding, connection,
-                                    batch, retryWrites, callback);
+                    try {
+                        if (isSecondAttempt || !shouldAttemptToRetryWrite(retryWrites, t)) {
+                            if (retryWrites && !isSecondAttempt) {
+                                logUnableToRetry(batch.getPayload().getPayloadType().toString(), t);
+                            }
+                            if (t instanceof MongoWriteConcernWithResponseException) {
+                                addBatchResult((BsonDocument) ((MongoWriteConcernWithResponseException) t).getResponse(), binding, connection,
+                                        batch, retryWrites, callback);
+                            } else {
+                                callback.onResult(null, t);
+                            }
                         } else {
-                            callback.onResult(null, t);
+                            retryExecuteBatchesAsync(binding, batch, t, callback.releaseConnectionAndGetWrapped());
                         }
-                    } else {
-                        retryExecuteBatchesAsync(binding, batch, t, callback.releaseConnectionAndGetWrapped());
+                    } catch (MongoClientException e) {
```

Comment:
I don't understand why this requires a new catch block.  Can you explain what the flow looks like?  I would expect that the same execution flow that handled the exception that is being wrapped would handle the MongoClientException.

Reply:
`shouldAttemptToRetryWrite` calls `CommandOperationHelper#isRetryableException` which throws our new MongoClientException. We do not catch the exception in `onResult` to send back to the callback.

Reply:
Resolved this as per our discussion.